### PR TITLE
Refactor build config to use convention plugins

### PR DESCRIPTION
**Pull request by aSemy**
Refactor build config to use convention plugins

Part 1 of #1 

- update .gitignore to include more IDE/OS/language files
- update to Gradle 8.2
- update Android Gradle Plugin to 8 (to match Gradle version)
- use Java 17 toolchain (required by Android 8)
- add centralised repo definitions
- create new build-logic included build for convention plugins
- move shared build logic into convention plugins
- re-enable Gradle parallel
- enabled [type-safe project accessors](https://docs.gradle.org/current/userguide/declaring_dependencies.html#sec:type-safe-project-accessors)
- Enabled a stricter configuration cache flag https://docs.gradle.org/current/userguide/configuration_cache.html#config_cache:stable, since configuration cache is now stable and wider usage is looming on the horizon

### CODE REVIEW

**Comment by jurmous** [comment]
Thanks for the work! In a first quick scan it all looks nice! 
The project all evolved over the years with Kotlin/Gradle and I see quite some nicer ways of setting up a build in your PR.
Let me know when you are done so I can review it. (It will unfortunately not be today but surely in next few days.)

File: build-logic/src/main/kotlin/maryk/conventions/kotlin-multiplatform-native.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        //val linuxArm64Main by getting { dependsOn(nativeMain) }
+        //val linuxArm64Test by getting { dependsOn(nativeTest) }
+
+        // Windows - MinGW
+        //val mingwX64Main by getting { dependsOn(nativeMain) }
+        //val mingwX64Test by getting { dependsOn(nativeTest) }
+
+        // Apple - macOS
+        val darwinMain by creating { dependsOn(nativeMain) }
+        val darwinTest by creating { dependsOn(nativeTest) }
```
Comment (aSemy) [inline build-logic/src/main/kotlin/maryk/conventions/kotlin-multiplatform-native.gradle.kts:81]: TODO: add darwin into the tree diagram above

**Comment by aSemy** [comment]
My pleasure! It's a lot of changes, so take your time, and I'm happy to answer any questions.

At the moment I'm just trying to get the tests running...

A couple of things I noted:

* The linuxX64 and mingwX64 Kotlin Native targets aren't supported by `io.maryk.rocksdb:rocksdb-multiplatform` - would it possible to enable those targets?
* linuxArm64 is a newer Kotlin Native target that isn't supported by kotlinx-datetime yet

**Comment by jurmous** [comment]
> My pleasure! It's a lot of changes, so take your time, and I'm happy to answer any questions.

Thanks!

> At the moment I'm just trying to get the tests running...
> 
> A couple of things I noted:
> 
> * The linuxX64 and mingwX64 Kotlin Native targets aren't supported by `io.maryk.rocksdb:rocksdb-multiplatform` - would it possible to enable those targets?
> * linuxArm64 is a newer Kotlin Native target that isn't supported by kotlinx-datetime yet

I see it as quite difficult to fully support the linuxX64 and mingwX64 targets for RocksDB. RocksDB is written in C++ and Kotlin does not provide C++ interop. So we would need to fall back to the way more limited C compatibility layer of RocksDB to provide support. 
I did once try to do an implementation with the C RocksDB calls but found it too limited for the MarykDB use case. 

I could manage to make it work with Darwin/Apple targets thanks to an Objective-C library which wrapped RocksDB which I then expanded. I could then call RocksDB on those platforms thanks to the Kotlin Obj-C interop. As far as I am aware, Kotlin does not provide ObjC compilation for other platforms than the Apple platform.
Since both Linux and Windows can always run the JVM I feel that those native targets were less needed to run MarykDB there, unlike the more limited iOS/ipadOS platforms.

So I am still hoping for Kotlin to expand to a limited C++ interop so it is easier to support those targets. Or somebody with more C/C++ experience needs to expand the C api of RocksDB. Or do you see alternatives which I missed?

And if you have any further issues to run any of the tests, let me know!

**Comment by aSemy** [comment]
I suspect that Kotlin/C++ interop isn't going to come, or if it does then it's many years away. C++ is just too complex compared to C to be able to generate bindings. 

But that said, it doesn't appear to be that hard to make a C wrapper for C++... so long as you're good enough! I've been playing around with [zphysics](https://github.com/michal-z/zig-gamedev/tree/36f8823cb39e33357e0cf0e0cb743dcdb10f1264/libs/zphysics), which is a C wrapper for [Jolt](https://github.com/jrouwe/JoltPhysics), which is written in C++. The bindings do seem quite formulaic - so perhaps they could be auto-generated.

I actually tried really hard to write some wrappers for the RocksDB C API. [I got really far](https://github.com/aSemy/kotlin-on-the-rocksdb), but I gave up because of Kotlin Native issues. It's a shame the RocksDB C API doesn't match the same features as the C++ API though.

I'm keeping an eye on https://github.com/whyoleg/ffi-kotlin, which would hopefully help with re-using a C library across all Kotlin platforms, which could help justify the cost of writing C bindings.

> And if you have any further issues to run any of the tests, let me know!

I'm just getting a weird error `Unresolved reference: Test`. For some reason `api(kotlin("test"))` and `api(kotlin("test-common"))` and `api(kotlin("test-annotations"))` isn't working...

**Comment by aSemy** [comment]
Alright, I think I've completed most of the refactoring and now I'm stuck on a couple of issues.

1. The tests in `maryk.datastore.rocksdb.RocksDBDataStoreMigrationTest` and `maryk.datastore.rocksdb.RocksDBDataStoreTest` fail on iosX64
    
    ```
    maryk.rocksdb.RocksDBException: Operation couldn't be completed: IO error: No such file or directory: While mkdir if missing: ./build/test-database/no-history: No such file or directory
    maryk.rocksdb.RocksDBException: Operation couldn't be completed: IO error: No such file or directory: While mkdir if missing: ./build/test-database/no-history: No such file or directory
    ```

   I'm not sure if this is a problem specific to iosX64, or perhaps the build directory isn't being properly cleaned. I think it's unusual that the same build directory is used for all tests, and all targets. The test should probably be updated to use a temporary directory.

2. Something is wrong with the task `:store:rocksdb:linkDebugTestMacosArm64`, I get a big error:

    ```
    > Task :store:rocksdb:linkDebugTestMacosArm64 FAILED
    e: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld invocation reported errors
    The /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld command returned non-zero exit code: 1.
    output:
    Undefined symbols for architecture arm64:
      "_objc_msgSend$median", referenced from:
          -[RocksDBStatisticsHistogram description] in libobjectiveRocks-macOS.a(RocksDBStatisticsHistogram.o)
      "_objc_msgSend$percentile99", referenced from:
          -[RocksDBStatisticsHistogram description] in libobjectiveRocks-macOS.a(RocksDBStatisticsHistogram.o)
      "_objc_msgSend$average", referenced from:
          -[RocksDBStatisticsHistogram description] in libobjectiveRocks-macOS.a(RocksDBStatisticsHistogram.o)
      "_objc_msgSend$min", referenced from:
          -[RocksDBStatisticsHistogram description] in libobjectiveRocks-macOS.a(RocksDBStatisticsHistogram.o)
      "_objc_msgSend$max", referenced from:
          -[RocksDBStatisticsHistogram description] in libobjectiveRocks-macOS.a(RocksDBStatisticsHistogram.o)
      "_objc_msgSend$setDb:", referenced from:
          -[RocksDBSnapshot initWithSnapshot:db:] in libobjectiveRocks-macOS.a(RocksDBSnapshot.o)
      "_objc_msgSend$db", referenced from:
          -[RocksDBSnapshot close] in libobjectiveRocks-macOS.a(RocksDBSnapshot.o)
      "_objc_msgSend$setSnapshot:", referenced from:
    ...
    ```

   Do you have any idea what the problem might be?

**Pull request merged**